[Content Filtering] Make ParentalControlsContentFilter::finishedAddingData wait for decision before returning

szewai opened the pull request:
#### 26b4f3b13848a3be71a83cc3212e84a1c26ff61d
<pre>
[Content Filtering] Make ParentalControlsContentFilter::finishedAddingData wait for decision before returning
<a href="https://bugs.webkit.org/show_bug.cgi?id=292566">https://bugs.webkit.org/show_bug.cgi?id=292566</a>
<a href="https://rdar.apple.com/145777952">rdar://145777952</a>

Reviewed by Per Arne Vollan.

Clients expect that filtering result is immediately available after finishedAddingData is called, see
ContentFilter::continueAfterNotifyFinished for an example. In existing implementation, there is no guarantee for that
when using WebContentRestrictions framework for filtering, since the WCRBrowserEngineClient SPI is async. To implement
that, this patches adopts the SPI that takes a queue parameter, so that the result can be set on background queue and
ParentalControlsContentFilter::finishedAddingData can be blocked waiting for the result on the main thread.

* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::forEachContentFilterUntilBlocked):
(WebCore::ContentFilter::didDecide):
(WebCore::ContentFilter::handleProvisionalLoadFailure):
* Source/WebCore/loader/ContentFilter.h:
* Source/WebCore/platform/PlatformContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::create):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::globalQueue):
(WebCore::ParentalControlsContentFilter::create):
(WebCore::ParentalControlsContentFilter::responseReceived):
(WebCore::ParentalControlsContentFilter::finishedAddingData):
(WebCore::ParentalControlsContentFilter::didReceiveAllowDecisionOnQueue):
(WebCore::ParentalControlsContentFilter::updateFilterStateOnMain):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::ParentalControlsURLFilter::isURLAllowed):
* Source/WebCore/testing/MockContentFilter.cpp:
(WebCore::MockContentFilter::create):
* Source/WebCore/testing/MockContentFilter.h:

Canonical link: <a href="https://commits.webkit.org/294563@main">https://commits.webkit.org/294563@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a936974070270478b39017c862d72015e034e069

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102259 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21927 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12242 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107419 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52895 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30434 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77806 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34797 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105265 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17208 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92309 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58144 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17039 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10337 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52253 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86868 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10407 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109794 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29391 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21664 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86786 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88512 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86374 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21981 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31185 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8904 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23633 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29319 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34614 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29130 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32453 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30689 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -158,8 +169,18 @@ static inline bool canHandleResponse(const ResourceResponse& response)
+        if (m_state != State::Filtering)
+            return;
+
+        // Callers expect state is ready after finishing adding data.
+        Locker resultLocker { m_resultLock };
+        while (!m_isAllowdByWebContentRestrictions)
+            m_resultCondition.wait(m_resultLock);
+        m_state = *m_isAllowdByWebContentRestrictions ? State::Allowed : State::Blocked;
+        m_replacementData = std::exchange(m_webContentRestrictionsReplacementData, nullptr);
```
In a possible follow-up patch, would it be feasible to add a completion handler to `finishedAddingData`, so that we don't have to block the main thread? This does not block landing this PR, though.

szewai replied:
It's possible but would require some amount of refactoring (at all callers), so this seems a safer change right now. We should consider about it later, and I will file a bug about this.
Thanks for review!

pvollan reviewed (approved):
R=me.

szewai committed:
[Content Filtering] Make ParentalControlsContentFilter::finishedAddingData wait for decision before returning
https://bugs.webkit.org/show_bug.cgi?id=292566
rdar://145777952

Reviewed by Per Arne Vollan.

Clients expect that filtering result is immediately available after finishedAddingData is called, see
ContentFilter::continueAfterNotifyFinished for an example. In existing implementation, there is no guarantee for that
when using WebContentRestrictions framework for filtering, since the WCRBrowserEngineClient SPI is async. To implement
that, this patches adopts the SPI that takes a queue parameter, so that the result can be set on background queue and
ParentalControlsContentFilter::finishedAddingData can be blocked waiting for the result on the main thread.

* Source/WebCore/loader/ContentFilter.cpp:
(WebCore::ContentFilter::forEachContentFilterUntilBlocked):
(WebCore::ContentFilter::didDecide):
(WebCore::ContentFilter::handleProvisionalLoadFailure):
* Source/WebCore/loader/ContentFilter.h:
* Source/WebCore/platform/PlatformContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.mm:
(WebCore::NetworkExtensionContentFilter::create):
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
(WebCore::globalQueue):
(WebCore::ParentalControlsContentFilter::create):
(WebCore::ParentalControlsContentFilter::responseReceived):
(WebCore::ParentalControlsContentFilter::finishedAddingData):
(WebCore::ParentalControlsContentFilter::didReceiveAllowDecisionOnQueue):
(WebCore::ParentalControlsContentFilter::updateFilterStateOnMain):
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.h:
* Source/WebCore/platform/cocoa/ParentalControlsURLFilter.mm:
(WebCore::ParentalControlsURLFilter::isURLAllowed):
* Source/WebCore/testing/MockContentFilter.cpp:
(WebCore::MockContentFilter::create):
* Source/WebCore/testing/MockContentFilter.h:

Canonical link: https://commits.webkit.org/294563@main
